topology.go

```diff
@@ -272,7 +276,87 @@ func (gs *generationSourcePre4_4) getGeneration(genTime time.Time, consistency g
 }
 
 func (gs *generationSourcePre4_4) getGenerationTimes(consistency gocql.Consistency) ([]time.Time, error) {
-	iter := gs.session.Query("SELECT time FROM " + generationsTableNamePre4_4).
+	iter := gs.session.Query("SELECT time FROM " + systemDistributedKeyspace + "." + generationsTableNamePre4_4).
+		Consistency(consistency).
+		Iter()
+	var (
+		times    []time.Time
+		currTime time.Time
+	)
+	for iter.Scan(&currTime) {
+		times = append(times, currTime)
+	}
+	if err := iter.Close(); err != nil {
+		return nil, err
+	}
+	return times, nil
+}
+
+// Follows the migration procedure from Scylla's documentation
+// https://docs.scylladb.com/using-scylla/cdc/cdc-querying-streams/
+func (gs *generationSourcePre4_4) maybeUpgrade(consistency gocql.Consistency) (generationSource, error) {
+	meta, err := gs.session.KeyspaceMetadata(systemDistributedKeyspace)
+	if err != nil {
+		return gs, err
+	}
+
+	// Check if table is present
+	_, ok := meta.Tables[streamsTableSince4_4]
+	if !ok {
+		// Don't upgrade, the new table is not there yet
+		return gs, nil
+	}
+
+	// Was the migration completed?
+	data := make(map[string]interface{})
+	err = gs.session.Query("SELECT streams_timestamp FROM system.cdc_local WHERE key = 'rewritten'").
+		Consistency(consistency).
```

Comment:
you're querying the local tables with QUORUM/ALL? Does that even work?

Reply:
Im sure it works for CL=ALL (where by "works" I mean the query is accepted by Scylla) because it was covered in the test I did (the cluster had 3 nodes during the rolling upgrade). The consistency is probably ignored when querying local tables.

I'll drop explicit setting of consistency here anyway, because it doesn't have much sense for local tables.

Reply:
Done